packages/permission-controller/src/rpc-methods/revokePermissions.ts

```diff
@@ -0,0 +1,80 @@
+import type { JsonRpcEngineEndCallback } from '@metamask/json-rpc-engine';
+import type {
+  Json,
+  JsonRpcRequest,
+  NonEmptyArray,
+  PendingJsonRpcResponse,
+} from '@metamask/utils';
+
+import { invalidParams } from '../errors';
+import type { PermissionConstraint } from '../Permission';
+import type { PermittedHandlerExport } from '../utils';
+import { MethodNames } from '../utils';
+
+export const revokePermissionsHandler: PermittedHandlerExport<
+  RevokePermissionsHooks,
+  RevokePermissionArgs,
+  null
+> = {
+  methodNames: [MethodNames.requestPermissions],
+  implementation: revokePermissionsImplementation,
+  hookNames: {
+    revokePermissionsForOrigin: true,
+  },
+};
+
+type RevokePermissionArgs = Record<
+  PermissionConstraint['parentCapability'],
+  Json
+>;
+
+type RevokePermissions = (
+  permissions: NonEmptyArray<PermissionConstraint['parentCapability']>,
+) => void;
+
+export type RevokePermissionsHooks = {
+  revokePermissionsForOrigin: RevokePermissions;
+};
+
+/**
+ * Revoke Permissions implementation to be used in JsonRpcEngine middleware.
+ *
+ * @param req - The JsonRpcEngine request
+ * @param res - The JsonRpcEngine result object
+ * @param _next - JsonRpcEngine next() callback - unused
+ * @param end - JsonRpcEngine end() callback
+ * @param options - Method hooks passed to the method implementation
+ * @param options.revokePermissionsForOrigin - A hook that revokes given permission keys for an origin
+ * @returns A promise that resolves to nothing
+ */
+async function revokePermissionsImplementation(
+  req: JsonRpcRequest<RevokePermissionArgs>,
+  res: PendingJsonRpcResponse<null>,
+  _next: unknown,
+  end: JsonRpcEngineEndCallback,
+  { revokePermissionsForOrigin }: RevokePermissionsHooks,
+): Promise<void> {
+  const { params } = req;
+
+  const param = params?.[0];
+
+  if (!param) {
+    return end(invalidParams({ data: { request: req } }));
+  }
+
+  // For now, this API revokes the entire permission key
+  // even if caveats are specified.
+  const permissionKeys = Object.keys(param);
+
+  if (permissionKeys.length === 0) {
+    return end(invalidParams({ data: { request: req } }));
+  }
+
+  revokePermissionsForOrigin(
+    permissionKeys as NonEmptyArray<PermissionConstraint['parentCapability']>,
+  );
```

Comment:
Do we need pass along origin here? Or will the hook have that context somehow?

Reply:
Normally the hooks are bound to the requesting origin and thus not needed to be passed

Reply:
oh hmm looks like we pass it as an arg for many other hooks. see `setNetworkClientIdForDomain` and `handleWatchAssetRequest` for example

Reply:
yeah ok we can do that. We should go and use this pattern for the other handlers as well.

Reply:
It depends on how the RPC method is implemented. The origin cannot currently be bound in a restricted RPC method handler for instance.